base/compiler/tfuncs.jl

```diff
@@ -815,17 +816,24 @@ function fieldtype_nothrow(@nospecialize(s0), @nospecialize(name))
 
     su = unwrap_unionall(s0)
     if isa(su, Union)
-        return fieldtype_nothrow(rewrap_unionall(su.a, s0), name) && fieldtype_nothrow(rewrap_unionall(su.b, s0), name)
+        return fieldtype_nothrow(rewrap_unionall(su.a, s0), name) &&
+               fieldtype_nothrow(rewrap_unionall(su.b, s0), name)
     end
 
     s = instanceof_tfunc(s0)[1]
-    u = unwrap_unionall(s)
-    return _fieldtype_nothrow(u, name)
+    s === Bottom && return false # always
+    return _fieldtype_nothrow(s, name)
 end
 
-function _fieldtype_nothrow(@nospecialize(u), name::Const)
+function _fieldtype_nothrow(@nospecialize(s), name::Const)
+    u = unwrap_unionall(s)
     if isa(u, Union)
-        return _fieldtype_nothrow(u.a, name) || _fieldtype_nothrow(u.b, name)
+        return _fieldtype_nothrow(u.a, name) && _fieldtype_nothrow(u.b, name)
```

Comment:
This is overly conservative --- for example `fieldtype(Union{Int,Tuple{Any}}, 1)` doesn't throw.

Reply:
Ah, OK, I need to start propagating the "exact" flag from instanceof_tfunc then